File: packages/native/android/src/main/java/expo/modules/jig/JigDispatcher.java
package expo.modules.jig;

import android.os.Handler;
import android.os.Looper;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class JigDispatcher {
    private static final int PROTOCOL_VERSION = 1;
    private static final String SERVER_VERSION = "0.1.0";

    private final List<JigMiddleware> middlewares;
    private final Map<String, JigHandler> handlerMap;
    private final List<String> supportedDomains;
    private final List<String> commandNames;
    private final Handler mainHandler = new Handler(Looper.getMainLooper());

    public JigDispatcher(List<JigMiddleware> middlewares,
                         List<JigHandler> handlers,
                         List<String> supportedDomains) {
        this.middlewares = middlewares;
        this.supportedDomains = supportedDomains;

        this.handlerMap = new HashMap<>();
        this.commandNames = new ArrayList<>();
        for (JigHandler handler : handlers) {
            String method = handler.getMethod();
            if (handlerMap.containsKey(method)) {
                throw new IllegalArgumentException(
                    "Duplicate handler for method: " + method);
            }
            handlerMap.put(method, handler);
            if (!"client.hello".equals(method)) {
                commandNames.add(method);
            }
        }
    }

    /** Send server.hello to a newly connected client */
    public void handleOpen(JigSessionContext context) {
        try {
            JSONObject protocol = new JSONObject();
            protocol.put("name", "jig");
            protocol.put("version", PROTOCOL_VERSION);
            protocol.put("min", PROTOCOL_VERSION);
            protocol.put("max", PROTOCOL_VERSION);

            JSONObject params = new JSONObject();
            params.put("protocol", protocol);
            params.put("server", SERVER_VERSION);
            params.put("app", context.getAppInfo().toJSON());
            params.put("commands", new JSONArray(commandNames));
            params.put("domains", new JSONArray(supportedDomains));

            JSONObject message = new JSONObject();
            message.put("jsonrpc", "2.0");
            message.put("method", "server.hello");
            message.put("params", params);

            context.sendText(message.toString());
        } catch (JSONException e) {
            System.out.println("[Jig] Failed to build server.hello: "
                               + e.getMessage());
        }
    }

    /** Parse, validate, route, and respond to an incoming message */
    public void dispatch(String text, JigSessionContext context) {
        // 1. Parse JSON
        JSONObject json;
        try {
            json = new JSONObject(text);
        } catch (JSONException e) {
            sendParseError("Malformed JSON", context);
            return;
        }

        // 2. Extract method, id, params
        String jsonrpc = json.optString("jsonrpc", "");
        String method = json.has("method") ? json.optString("method") : null;
        if (!"2.0".equals(jsonrpc) || method == null) {
            Integer id = json.has("id") ? json.optInt("id") : null;
            sendError(JigErrors.INVALID_REQUEST,
                "Missing or invalid 'jsonrpc' or 'method' field",
                null, id, context);
            return;
        }

        final Integer msgId = json.has("id") ? json.optInt("id") : null;
        final JSONObject params = json.optJSONObject("params");
        final boolean isCommand = (msgId != null);

        // 3. Run middleware chain
        for (JigMiddleware middleware : middlewares) {
            try {
                middleware.validate(method, params, context);
            } catch (JigException e) {
                if (isCommand) {
                    sendError(e.getCode(), e.getMessage(),
                              e.getData(), msgId, context);
                } else {
                    System.out.println("[Jig] Middleware rejected notification '"
                                       + method + "': " + e.getMessage());
                }
                return;
            }
        }

        // 4. Look up handler
        JigHandler handler = handlerMap.get(method);
        if (handler == null) {
            if (isCommand) {
                sendError(JigErrors.METHOD_NOT_FOUND,
                    "Method '" + method + "' not found",
                    null, msgId, context);
            } else {
                System.out.println("[Jig] No handler for notification '"
                                   + method + "'");
            }
            return;
        }

        // 5. Dispatch to handler's declared thread
        final String finalMethod = method;
        Runnable execute = () -> {
            executeHandler(handler, params, context,
                           finalMethod, msgId, isCommand);
        };

        switch (handler.getThreadTarget()) {
            case WEBSOCKET:
                execute.run();
                break;
            case MAIN:
                mainHandler.post(execute);
                break;
        }
    }

    // --- Handler execution ---

    private void executeHandler(JigHandler handler, JSONObject params,
                                JigSessionContext context, String method,
                                Integer msgId, boolean isCommand) {
        try {
            Object result = handler.handle(params, context);

            // Special case: update session context after successful handshake
            if ("client.hello".equals(method) && result instanceof JSONObject) {
                JSONObject resultObj = (JSONObject) result;
                context.setSessionId(resultObj.optString("sessionId"));
                JSONObject negotiated = resultObj.optJSONObject("negotiated");
                if (negotiated != null) {
                    context.setNegotiatedVersion(
                        negotiated.optInt("protocol"));
                }
            }

            if (isCommand) {
                sendResult(result, msgId, context);
            }
        } catch (JigException e) {
            if (isCommand) {
                sendError(e.getCode(), e.getMessage(),
                          e.getData(), msgId, context);
            } else {
                System.out.println("[Jig] Handler error for '" + method
                                   + "': " + e.getMessage());
            }
        } catch (Exception e) {
            if (isCommand) {
                sendError(JigErrors.INTERNAL_ERROR,
                    "Internal error: " + e.getMessage(),
                    null, msgId, context);
            } else {
                System.out.println("[Jig] Internal error for '" + method
                                   + "': " + e.getMessage());
            }
        }
    }

    // --- Response formatting ---

    private void sendResult(Object result, int msgId,
                            JigSessionContext context) {
        try {
            JSONObject response = new JSONObject();
            response.put("jsonrpc", "2.0");
            response.put("id", msgId);
            response.put("result", result != null ? result : JSONObject.NULL);
            context.sendText(response.toString());
        } catch (JSONException e) {
            System.out.println("[Jig] Failed to format result: "
                               + e.getMessage());
        }
    }

    private void sendError(int code, String message, JSONObject data,
                           Integer msgId, JigSessionContext context) {
        try {
            JSONObject error = new JSONObject();
            error.put("code", code);
            error.put("message", message);
            if (data != null) {
                error.put("data", data);
            }

            JSONObject response = new JSONObject();
            response.put("jsonrpc", "2.0");
            response.put("id", msgId != null ? (Object) msgId : JSONObject.NULL);
            response.put("error", error);
            context.sendText(response.toString());
        } catch (JSONException e) {
            System.out.println("[Jig] Failed to format error: "
                               + e.getMessage());
        }
    }

    private void sendParseError(String message, JigSessionContext context) {
        try {
            JSONObject error = new JSONObject();
            error.put("code", JigErrors.PARSE_ERROR);
            error.put("message", message);

            JSONObject response = new JSONObject();
            response.put("jsonrpc", "2.0");
            response.put("id", JSONObject.NULL);
            response.put("error", error);
            context.sendText(response.toString());
        } catch (JSONException e) {
            System.out.println("[Jig] Failed to format parse error: "
                               + e.getMessage());
        }
    }
}

File: packages/native/ios/JigDispatcher.m
// ios/JigDispatcher.m

#import "JigDispatcher.h"
#import "JigHandler.h"
#import "JigMiddleware.h"
#import "JigSessionContext.h"
#import "JigAppInfo.h"
#import "JigErrors.h"

static const NSInteger kProtocolVersion = 1;
static NSString *const kServerVersion = @"0.1.0";

@interface JigDispatcher ()
@property (nonatomic, strong) NSArray<id<JigMiddleware>> *middlewares;
@property (nonatomic, strong) NSDictionary<NSString *, id<JigHandler>> *handlerMap;
@property (nonatomic, strong) NSArray<NSString *> *supportedDomains;
@property (nonatomic, strong) NSArray<NSString *> *commandNames;
@property (nonatomic, strong) dispatch_queue_t queue;
@end

@implementation JigDispatcher

- (instancetype)initWithMiddlewares:(NSArray<id<JigMiddleware>> *)middlewares
                           handlers:(NSArray<id<JigHandler>> *)handlers
                   supportedDomains:(NSArray<NSString *> *)supportedDomains
                              queue:(dispatch_queue_t)queue {
    self = [super init];
    if (self) {
        _middlewares = middlewares;
        _supportedDomains = supportedDomains;
        _queue = queue;

        NSMutableDictionary *map = [NSMutableDictionary new];
        NSMutableArray *commands = [NSMutableArray new];
        for (id<JigHandler> handler in handlers) {
            NSString *method = handler.method;
            NSAssert(map[method] == nil,
                     @"[Jig] Duplicate handler for method: %@", method);
            map[method] = handler;
            if (![method isEqualToString:@"client.hello"]) {
                [commands addObject:method];
            }
        }
        _handlerMap = [map copy];
        _commandNames = [commands copy];
    }
    return self;
}

#pragma mark - Connection lifecycle

- (void)handleOpen:(JigSessionContext *)context {
    NSDictionary *params = @{
        @"protocol": @{
            @"name": @"jig",
            @"version": @(kProtocolVersion),
            @"min": @(kProtocolVersion),
            @"max": @(kProtocolVersion),
        },
        @"server": kServerVersion,
        @"app": [context.appInfo toDictionary],
        @"commands": self.commandNames,
        @"domains": self.supportedDomains,
    };

    NSDictionary *message = @{
        @"jsonrpc": @"2.0",
        @"method": @"server.hello",
        @"params": params,
    };

    context.sendText([self jsonStringFromDictionary:message]);
}

#pragma mark - Dispatch

- (void)dispatch:(NSString *)text context:(JigSessionContext *)context {
    // 1. Parse JSON
    NSData *data = [text dataUsingEncoding:NSUTF8StringEncoding];
    if (!data) {
        [self sendParseError:@"Invalid UTF-8" context:context];
        return;
    }

    NSError *parseErr = nil;
    id parsed = [NSJSONSerialization JSONObjectWithData:data options:0 error:&parseErr];
    if (parseErr || ![parsed isKindOfClass:[NSDictionary class]]) {
        [self sendParseError:@"Malformed JSON" context:context];
        return;
    }
    NSDictionary *json = (NSDictionary *)parsed;

    // 2. Extract method, id, params
    NSString *jsonrpc = json[@"jsonrpc"];
    NSString *method = json[@"method"];
    if (![jsonrpc isEqualToString:@"2.0"] || ![method isKindOfClass:[NSString class]]) {
        [self sendErrorWithCode:JigErrorCodeInvalidRequest
                        message:@"Missing or invalid 'jsonrpc' or 'method' field"
                           data:nil
                             id:json[@"id"]
                        context:context];
        return;
    }

    NSNumber *msgId = json[@"id"];  // nil for notifications
    NSDictionary *params = json[@"params"];
    BOOL isCommand = (msgId != nil);

    // 3. Run middleware chain
    for (id<JigMiddleware> middleware in self.middlewares) {
        NSError *mwError = nil;
        BOOL ok = [middleware validateMethod:method
                                     params:params
                                    context:context
                                      error:&mwError];
        if (!ok && mwError) {
            if (isCommand) {
                [self sendNSError:mwError id:msgId context:context];
            } else {
                NSLog(@"[Jig] Middleware rejected notification '%@': %@",
                      method, mwError.localizedDescription);
            }
            return;
        }
    }

    // 4. Look up handler
    id<JigHandler> handler = self.handlerMap[method];
    if (!handler) {
        if (isCommand) {
            [self sendErrorWithCode:JigErrorCodeMethodNotFound
                            message:[NSString stringWithFormat:
                                     @"Method '%@' not found", method]
                               data:nil
                                 id:msgId
                            context:context];
        } else {
            NSLog(@"[Jig] No handler for notification '%@'", method);
        }
        return;
    }

    // 5. Dispatch to handler's declared thread
    switch (handler.threadTarget) {
        case JigThreadTargetWebSocket:
            [self executeHandler:handler
                          params:params
                         context:context
                          method:method
                           msgId:msgId
                       isCommand:isCommand];
            break;

        case JigThreadTargetMain:
            dispatch_async(dispatch_get_main_queue(), ^{
                [self executeHandler:handler
                              params:params
                             context:context
                              method:method
                               msgId:msgId
                           isCommand:isCommand];
            });
            break;
    }
}

#pragma mark - Handler execution

- (void)executeHandler:(id<JigHandler>)handler
                params:(NSDictionary *)params
               context:(JigSessionContext *)context
                method:(NSString *)method
                 msgId:(NSNumber *)msgId
             isCommand:(BOOL)isCommand {

    NSError *handlerError = nil;
    id result = [handler handleParams:params context:context error:&handlerError];

    // Response must be sent on the WebSocket queue
    void (^respond)(void) = ^{
        if (handlerError) {
            if (isCommand) {
                [self sendNSError:handlerError id:msgId context:context];
            } else {
                NSLog(@"[Jig] Handler error for '%@': %@",
                      method, handlerError.localizedDescription);
            }
            return;
        }

        // Special case: update session context after successful handshake
        if ([method isEqualToString:@"client.hello"] &&
            [result isKindOfClass:[NSDictionary class]]) {
            NSDictionary *resultDict = (NSDictionary *)result;
            context.sessionId = resultDict[@"sessionId"];
            context.negotiatedVersion =
                [resultDict[@"negotiated"][@"protocol"] integerValue];
        }

        if (isCommand) {
            [self sendResult:result id:msgId context:context];
        }
    };

    if (handler.threadTarget == JigThreadTargetWebSocket) {
        respond();
    } else {
        dispatch_async(self.queue, respond);
    }
}

#pragma mark - Response formatting

- (void)sendResult:(id)result
                id:(NSNumber *)msgId
           context:(JigSessionContext *)context {
    NSDictionary *response = @{
        @"jsonrpc": @"2.0",
        @"id": msgId,
        @"result": result ?: [NSNull null],
    };
    context.sendText([self jsonStringFromDictionary:response]);
}

- (void)sendNSError:(NSError *)error
                 id:(NSNumber *)msgId
            context:(JigSessionContext *)context {
    NSDictionary *errorData = error.userInfo[JigErrorDataKey];
    [self sendErrorWithCode:error.code
                    message:error.localizedDescription
                       data:errorData
                         id:msgId
                    context:context];
}

- (void)sendErrorWithCode:(NSInteger)code
                  message:(NSString *)message
                     data:(NSDictionary *)data
                       id:(NSNumber *)msgId
                  context:(JigSessionContext *)context {
    // Use NSNull for id if nil (parse errors / invalid requests)
    id responseId = msgId ?: [NSNull null];

    NSMutableDictionary *errorDict = [@{
        @"code": @(code),
        @"message": message,
    } mutableCopy];
    if (data) {
        errorDict[@"data"] = data;
    }

    NSDictionary *response = @{
        @"jsonrpc": @"2.0",
        @"id": responseId,
        @"error": [errorDict copy],
    };
    context.sendText([self jsonStringFromDictionary:response]);
}

- (void)sendParseError:(NSString *)message
               context:(JigSessionContext *)context {
    [self sendErrorWithCode:JigErrorCodeParseError
                    message:message
                       data:nil
                         id:nil
                    context:context];
}

#pragma mark - JSON serialization

- (NSString *)jsonStringFromDictionary:(NSDictionary *)dict {
    NSData *data = [NSJSONSerialization dataWithJSONObject:dict
                                                  options:NSJSONWritingSortedKeys
                                                    error:nil];
    return [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
}

@end
